[LSV] Enhance LoadStoreVectorizer to Handle Disjoint Flag in OR Instructions and Restore Vectorization Opportunities

LiHao217 opened the pull request:
### [LSV] Enable LoadStoreVectorizer to recognize disjoint ORs by updating SCEV analysis

### Summary:
This patch enhances the LoadStoreVectorizer (LSV) pass to handle the disjoint flag for OR instructions at the Scalar Evolution (SCEV) analysis level.  By doing so, we address a regression that caused a loss in vectorization opportunities when processing OR operations that are actually disjoint additive operations.

### Background:
A previous change ([PR#74467](https://github.com/llvm/llvm-project/pull/74467)) introduced the disjoint flag as a representation change for OR instructions with no common bits set, as identified by haveNoCommonBitsSet().  This was intended to streamline the identification process and make it easier to revert such inferences.  However, this led LSV to miss converting disjoint ORs into ADDs, causing a performance regression, particularly on AMD and NVIDIA architectures where such vectorizations are crucial.

### Modifications:
The regression is addressed by enhancing the SCEV analysis to detect the disjoint flag on OR instructions and consider them as candidates for addition operation transformations.   The change occurs within the MatchBinaryOp function, where we now set the disjoint flag on OR instructions after checking operands for common bits using haveNoCommonBitsSet().   By shifting the focus from LSV to SCEV, we better leverage LLVM's existing infrastructure for handling such special cases.

SCEV's role in analyzing and simplifying expressions throughout the compiler optimization phases makes it a more strategic point for detecting disjoint OR instructions.

SCEV's inherent design to model and optimize scalar expressions allows us to integrate `or disjoint` detection seamlessly.    By integrating into SCEV’s `MatchBinaryOp` function, we empower LLVM to naturally apply this transformation during the expression matching phase.    This allows the subsequent vectorization passes, such as LSV, to benefit from the already identified and transformed instructions, ensuring that vectorization opportunities involving `or disjoint` are not missed.

The detection logic checks for disjointness of OR operands similar to what was previously done, but now the flagged instructions become part of SCEV's canonical form.    This not only fixes the regression introduced by the disjoint flag changes but also bolsters LLVM's ability to optimize such patterns consistently across different optimization passes.

### Original IR:
%linear_index3 = or i32 %linear_index_plus_base.fr, 3
%linear_index2 = or i32 %linear_index_plus_base.fr, 2
%linear_index1 = or i32 %linear_index_plus_base.fr, 1


### Transformed IR with disjoint flag:
%linear_index3 = or disjoint i32 %linear_index_plus_base.fr, 3
%linear_index2 = or disjoint i32 %linear_index_plus_base.fr, 2
%linear_index1 = or disjoint i32 %linear_index_plus_base.fr, 1


With the above transformation in place, LSV can identify these disjoint OR operations as additive, restoring its previously lost vectorization capabilities.

### Test Plan:
A new phase-ordering test has been added to demonstrate that without this patch, OR instructions lacking the disjoint flag were not vectorized under -O1, -O2, and -O3 optimization levels.  With the patch applied, SCEV successfully flags the OR instructions as disjoint, allowing LSV to vectorize them, as confirmed by the test.


Signed-off-by: lihao217 <lihao217@kuaishou.com>, Compiler Optimization Engineer at Kuaishou Technology

LiHao217 committed:
[LSV] Restore vectorization opportunities by handling disjoint flag in OR instructions

Prior to patch (https://github.com/llvm/llvm-project/pull/74467), LLVM used the haveNoCommonBitsSet() function to determine whether OR instructions had no common bits. If they didn't, the OR instructions were transformed into ADD instructions, allowing the LoadStoreVectorizer (LSV) to vectorize them.

After the patch was applied, the implementation switched to using a disjoint flag to indicate this condition. However, these OR instructions were no longer recognized as disjoint and were not converted into ADD instructions. This oversight led to a regression where LSV failed to vectorize these instructions.

This commit addresses the regression by updating LSV to identify if the OR instructions are disjoint, and if so, to transform the IR as follows to continue vectorization:

Original IR:

    %linear_index3 = or i32 %linear_index_plus_base.fr, 3
    %linear_index2 = or i32 %linear_index_plus_base.fr, 2
    %linear_index1 = or i32 %linear_index_plus_base.fr, 1

Transformed IR:

    %linear_index3 = or disjoint i32 %linear_index_plus_base.fr, 3
    %linear_index2 = or disjoint i32 %linear_index_plus_base.fr, 2
    %linear_index1 = or disjoint i32 %linear_index_plus_base.fr, 1

This modification will enable LSV to recognize these instructions and restore vectorization opportunities.

dtcxzyw commented on llvm/lib/Transforms/Vectorize/LoadStoreVectorizer.cpp:
```diff
@@ -437,9 +437,18 @@ bool Vectorizer::run() {
+        Value *Op1 = OrInst->getOperand(1);
+
+        if (haveNoCommonBitsSet(Op0, Op1, DL)) {
+          OrInst->setIsDisjoint(true);
```
Isn't it handled by `InstCombinerImpl::SimplifyDemandedUseBits`?

LiHao217 replied:
The backend I tested was not X86 but NVPTX. Without this change, llvm would not be able to autovectorize the following test case

```
Test case:
LLVM/test/Transforms/LoadStoreVectorizer/NVPTX/disjoint_or_vectorizer ll
Test the command:
opt -mtriple=nvptx64-nvidia-cuda -debug-pass-manager -passes=load-store-vectorizer -S -o -  llvm/test/Transforms/LoadStoreVectorizer/NVPTX/disjoint_or_vectorizer.ll
```

nikic reviewed (changes requested):
This needs a phase ordering test to demonstrate why the disjoint flag is not present at the point where LSV runs. Then we can either try to make sure the flag is present at that point, or make SCEV fall back to haveNoCommonBitsSet() -- I don't think that we want to infer it in LSV specifically.

LiHao217 commented:
> This needs a phase ordering test to demonstrate why the disjoint flag is not present at the point where LSV runs. Then we can either try to make sure the flag is present at that point, or make SCEV fall back to haveNoCommonBitsSet() -- I don't think that we want to infer it in LSV specifically.
@nikic 
Hi nikic,
I have provided a phase ordering test case that clearly demonstrates the scenario where the disjoint flag is missing when the LoadStoreVectorizer (LSV) runs. You can find the test case here:
[LLVM/test/Transforms/LoadStoreVectorizer/NVPTX/disjoint_or_vectorizer.ll](https://github.com/llvm/llvm-project/blob/ece0efcb68ac0868d128b70b92b83a5b1801b62e/llvm/test/Transforms/LoadStoreVectorizer/NVPTX/disjoint_or_vectorizer.ll)

You can run the test with the following command:
`opt -mtriple=nvptx64-nvidia-cuda -debug-pass-manager -passes=load-store-vectorizer -S -o - llvm/test/Transforms/LoadStoreVectorizer/NVPTX/disjoint_or_vectorizer.ll`

This test case demonstrates the issue with the missing disjoint flag when LSV runs, which helps in understanding why the disjoint flag is not being properly propagated through the prior optimization phases.

Could you please advise on the appropriate phase or location to set the disjoint flag? I am willing to make the necessary modifications based on your guidance.

Thank you for your help!

Best regards,
LiHao217

arsenm commented:
> have provided a phase ordering test case that clearly demonstrates the scenario where the disjoint flag is missing when the LoadStoreVectorizer (LSV) runs. You can find the test case here:

This is not a phase ordering test. It just runs a single pass. A phase ordering test would go in test/Transforms/PhaseOrdering/ and run opt -O1/-O2/-O3, not a single pass

LiHao217 commented:
> This is not a phase ordering test. It just runs a single pass. A phase ordering test would go in test/Transforms/PhaseOrdering/ and run opt -O1/-O2/-O3, not a single pass

@arsenm Thanks for the guidance, I will supplement the test. I am also rethinking about the position where this optimization should be implemented, and I will find a suitable position for it. Thank you

LiHao217 committed:
Merge branch 'main' into disjoint_or_vectorizer

LiHao217 committed:
[LSV] Fix regression in OR instruction vectorization

Summary:
   This commit resolves a regression in the LoadStoreVectorizer (LSV)
   due to a previous change in how OR instructions were handled. The
   regression prevented LSV from recognizing and vectorizing OR
   instructions marked as disjoint.

Background:
    A historical patch ([https://github.com/llvm/llvm-project/pull/74467](https://github.com/llvm/llvm-project/pull/74467))
    shifted the identification of disjoint OR instructions from using
    the haveNoCommonBitsSet() function to a disjoint flag within the
    instruction. However, this change inadvertently led to missed
    vectorization opportunities because LSV was no longer converting
    disjoint OR instructions to ADD instructions, as it did and not
    recognize the new flag.

    This patch updates the Scalar Evolution (SCEV) analysis to identify
    disjoint OR instructions based on the disjoint flag. By doing this,
    the instructions will be correctly interpreted and the LSV can restore
    lost vectorization opportunities.

Test plan:
    Added a new phase-ordering test that confirms the regression at -O1,
    -O2, and -O3 optimization levels, where the original OR instructions
    were not identified as disjoint and thus not vectorized. The test
    illustrates that with the new SCEV analysis, the transformed IR will
    properly have disjoint OR instructions vectorized by LSV.

Details:
    The modified test disjoint_or_vectorizer.ll showcases the restored
    vectorization with the following transformed IR:

Original IR:
    %linear_index3 = or i32 %linear_index_plus_base.fr, 3
    %linear_index2 = or i32 %linear_index_plus_base, 2
    %linear_index1 = or i32 %linear_index_plus_base, 1

Transformed IR with disjoint flag:
    %linear_index3 = or disjoint i32 %linear_index_plus_base, 3
    %linear_index2 = or disjoint i32 %linear_index_plus_base, 2
    %linear_index1 = or disjoint i32 %linear_index_plus_base, 1

With this change, we ensure the SCEV analysis can correctly handle and
optimize disjoint OR instructions.

Signed-off-by: lihao217 <lihao217@kuaishou.com>

LiHao217 commented:
> A new phase-ordering test has been added to demonstrate that without this patch, OR instructions lacking the disjoint flag were not vectorized under -O1, -O2, and -O3 optimization levels. With the patch applied, SCEV successfully flags the OR instructions as disjoint, allowing LSV to vectorize them, as confirmed by the test.

Hi @arsenm @nikic @dtcxzyw ,

Thank your for your review and feedback.  I have updated my patch and added a new test case: llvm/test/Transforms/PhaseOrdering/disjoint_or_vectorizer.ll.  This test case illustrates that prior to applying my patch, OR instructions in the IR were not flagged as disjoint at the -O1, -O2, and -O3 optimization levels.

I have also refined the implementation based on our discussion.  The detection of or disjoint has now been moved from the LoadStoreVectorizer (LSV) to the Scalar Evolution (SCEV) analysis.  This change provides a more appropriate and robust handling of OR instructions with the disjoint flag within the existing LLVM infrastructure.

The new patch ensures that OR instructions marked as disjoint are identified early in the optimization process during SCEV analysis, paving the way for proper vectorization in subsequent stages.

I believe this addresses the previous concerns and should make the vectorization logic more consistent.  Could you please take another look at the updated patch and the added test case?  Your feedback is greatly appreciated.

Best regards,
Hao LI

nikic commented on llvm/lib/Analysis/ScalarEvolution.cpp:
```diff
@@ -5290,6 +5290,15 @@ static std::optional<BinaryOp> MatchBinaryOp(Value *V, const DataLayout &DL,
+      Value *Op1 = OrInst->getOperand(1);
+
+      if (haveNoCommonBitsSet(Op0, Op1, DL)) {
+        OrInst->setIsDisjoint(true);
```
You cannot modify instructions inside SCEV. What you would do is add an `|| haveNoCommonBits` check in the if below.

LiHao217 replied:
Do you mean change it like this? I'm going to test it now：
`  case Instruction::Or: {
    if (auto *OrInst = dyn_cast<PossiblyDisjointInst>(Op)) {
      // Convert or disjoint into add nuw nsw.
      if (OrInst->isDisjoint() ||
          haveNoCommonBitsSet(OrInst->getOperand(0), OrInst->getOperand(1), DL)) {
        return BinaryOp(Instruction::Add, OrInst->getOperand(0), OrInst->getOperand(1),
                        /*IsNSW=*/true, /*IsNUW=*/true);
      }
    }
    return BinaryOp(Op);
  }`

nikic commented on llvm/test/Transforms/PhaseOrdering/disjoint_or_vectorizer.ll:
Is it possible to further reduce this? It's hard to understand test cases with so many instructions.

I can see that `or disjoint` is indeed missing -- but only if `-mtriple=nvptx64-nvidia-cuda` is used. Why does adding the nvptx triple break the disjoint inference

LiHao217 replied:
This is my mistake, the issue is not related to the -mtriple=nvptx64-nvidia-cuda option, and it can be removed. The nvptx triple does not break the disjoint inference.

LiHao217 replied:
You are right, the or disjoint is missing when -mtriple=nvptx64-nvidia-cuda is enabled, I will check this issue。
I found the reason is that the https://github.com/llvm/llvm-project/commit/cd564c6b4638e654edb68acf7f7d985c08edf468 this commit mention LSV optimization is not mature, So the default is to turn this optimization off, so I think for now I can make some small changes on LSV and then continue to let the user choose whether to enable LSV optimization or not.

nikic replied:
Did you figure out why the disjoint is only missing with the nvptx triple? I think we need to understand this before moving forward with the patch.

Having a smaller test case would help analyze this.

LiHao217 replied:
I am analyzing the problem, please wait, I will fix the failure of CI

LiHao217 committed:
[ScalarEvolution] Optimize OR instruction when operands are disjoint

Summary:
When handling the OR instruction within scalar evolution analysis, if it
is determined that the two operands have no common bits set (they are disjoint),
the OR instruction can be optimized as an ADD instruction with no signed wrap
(NSW) and no unsigned wrap (NUW) flags set.

This patch adjusts the existing logic to check for disjoint operands without
modifying the instructions themselves, aligning with the requirement to not
modify instructions inside the Scalar Evolution (SCEV) analysis. The
`isDisjoint` method along with `haveNoCommonBitsSet` function are utilized to
make this determination and then appropriately construct a BinaryOp representing
the optimized ADD instruction.

Reviewers: nikic

Subscribers: lihao217 lihao217@kuaishou.com, Compiler Optimization Engineer at Kuaishou Technology
